Redone #4189, working preview scrolling

GNUSheep opened the pull request:
Finally redone preview scrolling on current master. Works pretty fine.

I used code from: @Manosmer, and make it work with current version of helix.

Connects #1614

thomasaarholt commented:
Current keybindings:

> * `alt-j` / `shift-down` - Move preview down `scroll-lines` at a time
> * `alt-k` / `shift-up` - Move preview up `scroll-lines` at a time
> * `alt-d` - Move preview down half height of the preview area
> * `alt-u` - Move preview up half height of the preview area
> * `alt-f` / `alt-PageDown` - Move preview down one full preview area height
> * `alt-b` / `alt-PageUp` - Move preview up one full preview area height
> 
> _Originally posted by @EpocSquadron in https://github.com/helix-editor/helix/issues/4189#issuecomment-1279984697_
>

thomasaarholt commented:
I do observe that when scrolling, the syntax highlighting does not update. Only the text first shown in the preview pane is syntax highlighted.

<img width="1822" alt="Screenshot 2024-08-18 at 19 15 50" src="https://github.com/user-attachments/assets/ccab63ca-bd42-4fe6-a106-2ccd8fb0546a">

GNUSheep commented:
@thomasaarholt 

Hi! Thank you for your help. I fixed my mistakes, and now all should work well.

My only concern is about keybinding alt-PageUp, because as mentioned before:

> I'm not 100% certain alt- is valid with PageUp/PageDown in terminals, but worth a try.

> I tried the above on iTerm2 without success. Therefore, I did not include those keybindings. It might be a limitation just on my platform though.

In my case alt-PageDown works pretty fine, but alt-PageUp doesn't.
I tested on alattracity, and gnome terminal and got same result.

But as mentioned above it might be just a limitation on my platform.

GNUSheep commented:
@archseer 
@the-mikedavis

EmrysMyrddin commented:
I've tried it, works very well, very welcomed addition ! I hope it will be merged soon :-)

EmrysMyrddin commented:
I'm using Helix based on your PR, and I came across another issue:
<img width="1728" alt="image" src="https://github.com/user-attachments/assets/9101b8af-0fdd-4186-a9c5-7438626aca25">

Here the diagnostic is miss-rendered. It should not underline the whole comment.

Here how it looks in the main buffer:
<img width="1728" alt="image" src="https://github.com/user-attachments/assets/45d0923e-2f0a-4f6b-87d0-ec3ae99f877f">

Note: the diagnostic is rendered in the preview only if the file have been opened at least once since Helix started.

GNUSheep commented:
@EmrysMyrddin 

Hi! Thank u for your using helix based on my PR. 
I fixed that bug, 
(I just forgot to change offset.anchor in doc_diagnostics_highlights)

![image](https://github.com/user-attachments/assets/3e7aeac3-8413-4ec3-bcc8-3b52e50ecdb5)
![image](https://github.com/user-attachments/assets/f1b23b7e-5ba0-42df-bed7-9d22aa1d4ce8)

If u find more bugs I will be happy to fix them :)

GNUSheep commented:
@David-Else 
Hello! Looking at the code PageUp and PageDown, moves up or down by a exactly one page. But I agree with you that it is not so useful, I think that many people just search by using search bar rather than scrolling up. 

Using just PageUp or PageDown seems like nice idea.

GNUSheep commented:
@archseer @the-mikedavis @pascalkuthe

qiu-x commented:
Are there any blockers before this can be merged?
@the-mikedavis

rockboynton commented:
I can consistently reproduce a panic when using this. (disclaimer, I do have other branches too merged into my fork too, but I can only reproduce this issue when this PR is included):

1. Open a repo with lsp (rust-analyzer in my case)
2. Open global symbol picker (<space>S)
3. Typy-type
4. The panic occurs sometimes while typing, sometimes when I hit <esc> to close, sometimes if I close, re-open the global symbol picker and type more. 

Can you try to reproduce? It may be possible it is just my unique setup, but like I said, I've only seen this when using this PR.
```
󱞪 RUST_BACKTRACE=full hx .
thread 'main' panicked at /nix/store/eeeeeeeeeeeeeeeeeeeeeeeeeeeeeeee-vendor-cargo-deps/c19b7c6f923b580ac259164a89f2577984ad5ab09ee9d583b888f934adbbe8d0/ropey-1.6.1/src/slice.rs:453:41:
called `Result::unwrap()` on an `Err` value: Line index out of bounds: line index 6457, Rope/RopeSlice line count 436
stack backtrace:
   0:     0x55555661be0b - <std::sys::backtrace::BacktraceLock::print::DisplayBacktrace as core::fmt::Display>::fmt::h1b9dad2a88e955ff
   1:     0x55555577884b - core::fmt::write::h4b5a1270214bc4a7
   2:     0x555556617b73 - std::io::Write::write_fmt::hd04af345a50c312d
   3:     0x55555661de2a - std::panicking::default_hook::{{closure}}::h96ab15e9936be7ed
   4:     0x55555661da77 - std::panicking::default_hook::h3cacb9c27561ad33
   5:     0x55555661e6a8 - std::panicking::rust_panic_with_hook::hfe205f6954b2c97b
   6:     0x55555661e397 - std::panicking::begin_panic_handler::{{closure}}::h6cb44b3a50f28c44
   7:     0x55555661c2d9 - std::sys::backtrace::__rust_end_short_backtrace::hf1c1f2a92799bb0e
   8:     0x55555661e024 - rust_begin_unwind
   9:     0x555555680593 - core::panicking::panic_fmt::h3d8fc78294164da7
  10:     0x555555680a26 - core::result::unwrap_failed::hfa79a499befff387
  11:     0x555556106c47 - <helix_term::ui::picker::Picker<I,D> as helix_term::compositor::Component>::render::h33dd85836441c517
  12:     0x555556377c03 - helix_term::application::Application::render::{{closure}}::h108963a17ac2b9f0
  13:     0x55555638c7b0 - hx::main_impl::{{closure}}::h1105d798dbc7e566
  14:     0x5555563898c8 - tokio::runtime::park::CachedParkThread::block_on::h390deff2c402d9ac
  15:     0x55555642a4de - tokio::runtime::context::runtime::enter_runtime::h6b594ce14328c083
  16:     0x555556439f46 - tokio::runtime::runtime::Runtime::block_on::h03ec0308555e9d53
  17:     0x5555563cdb73 - hx::main::h3debb9f5c1109bd7
  18:     0x5555563f4143 - std::sys::backtrace::__rust_begin_short_backtrace::ha704d94dc880694d
  19:     0x5555563fbb3d - std::rt::lang_start::{{closure}}::hf9a24b6ef71d3115
  20:     0x55555660e685 - std::rt::lang_start_internal::h5e7c81cecd7f0954
  21:     0x5555563cdc65 - main
  22:     0x7ffff7c2a1fc - __libc_start_call_main
  23:     0x7ffff7c2a2b9 - __libc_start_main_alias_2
  24:     0x5555556f3535 - _start
  25:                0x0 - <unknown>
```

GNUSheep commented:
@rockboynton 

Hi! Thank you for pointing out the error. This panic was indeed related to this PR. I missed it during testing - _perhaps I didn't run enough tests_ -  so I am sorry for that oversight.

Together with @qiu-x, we have fixed the issue. We switched to using offset_anchor instead of offset_vertical_offset to move the preview, and this resolved the problem. We ran several tests and did not encounter any more panics.

I hope everything works fine now. I look forward to your feedback! Please let me know if everything is working as expected
I really hope it does. 😊

Thank you once again!

nik-rev commented:
this PR  also closes https://github.com/helix-editor/helix/issues/1614

GNUSheep commented:
Currently not, but I will add this idea to my to-do list, and implement it.

(It is just a case to add moving preview as static command in 'helix-term/src/commands.rs')

vishalpaudel commented:
What is stopping from this being merged to master?

n6v26r commented:
Why aren't we getting this merged?

qiu-x committed:
feat: Redone preview scrolling from helix-editor#4189

Co-authored-by: GNUSheep <zydekpiotr26@gmail.com>
Co-authored-by: Manos Mertzianis <manosmertzianis@gmail.com>

EmrysMyrddin commented:
@GNUSheep Can you rebase your branch ? It has conflicts now ><'

GNUSheep commented:
@archseer @the-mikedavis @pascalkuthe @sudormrfbin @dead10ck

the-mikedavis reviewed (commented):
It seems like this doesn't work well with soft-wrap enabled. Setting `editor.soft-wrap.enable` to `true` also enables soft-wrap in the picker and when a file has long lines I can't get to the bottom in the preview with `A-d`

GNUSheep commented:
Hi @the-mikedavis
Thanks you for your review, I will apply your suggestions as soon as I can

sdi2200262 commented:
Rebased this onto current master and applied the review here in #15854: scrolling is by visual lines now, so the bottom of a soft-wrapped file is reachable, `Ctrl-u`/`Ctrl-d` stay list navigation with the preview on `PageUp`/`PageDown` and `Alt-j`/`Alt-k`, and I fixed an offset accumulation at both edges. Credited you as co-author since it builds on your commit, @GNUSheep. Happy to fold it back here instead if you're still on this.

GNUSheep commented:
Hi! @sdi2200262, happy to see someone picked this up 😄. 
I had to drop it due to a lack of time, so I'm keeping my fingers crossed for you. Good luck!

sdi2200262 commented:
Thanks! Added you to the co-authors on #15854, alongside @qiu-x who authored the commit. Cheers for getting it this far.
